FF153 CSS basic shape closest-/farthest-corner (#44827)
mdn/content: files/en-us/web/css/reference/values/basic-shape/circle/index.md

### Values

- `<shape-radius>`
  - : This may be a {{cssxref("length")}}, or a {{cssxref("percentage")}} or values `closest-side` and `farthest-side`.
  - : This may be a {{cssxref("length")}}, a {{cssxref("percentage")}}, or any of the following keyword values:
    - `closest-side`

>       - : Uses the length from the center of the shape to the closest side of the reference box. For circles, this is the closest side in any dimension.

Reviewer: My question here for all instances of "closest" and "farthest", here and on the ellipse page is - "closest to or farthest from what?". It looks like it means "closest to or farthest from the inline/block start of the shape", but is this explained anywhere?

Author: @chrisdavidmills You've commented changed text. Isn't this answered by the new text - from the centre of the circle/ellipse.

Reviewer: Huh, yeah, I was getting rather confused here ;-)

The updated content is definitely better.

      - : The radius is the length from the center of the circle to the farthest side of the reference box, so that the circle's boundary just touches that side.
    - `closest-corner`

>       - : The radius is the length from the center of the circle to the closest corner of the reference box, so that the circle passes through that corner.

Reviewer: ```suggestion
      - : The radius is the length from the center of the circle to the closest corner of the reference box, so that the circle's boundary passes through that corner.
```

Same for the other instances too.

    - `farthest-corner`

>       - : The radius is the length from the center of the circle to the farthest corner of the reference box, so that the circle passes through that corner.

Author: ```suggestion
      - : The radius is the length from the center of the circle to the farthest corner of the reference box, so that the circle's boundary passes through that corner.
```

{{EmbedLiveSample("circle", "", "300px")}}

### Radius keyword interactive example

This example allows you to test the effect of moving the center point of the circle when using each of the four `<shape-radius>` keywords.

> The code defines a reference box (`#refbox`) that contains a `.fill` element.

Reviewer: So the `#refbox` is the pink dashed outline box? And that's the box we are setting `clip-path` on, to clip the circle shape? 

I think it is a bit confusing because the explanation is quite far away from the live demo, so by the time you get there, you are not sure what's happening. Also, the background is supposed to be clipped to the circle shape and the shape of the box, but you can see the entire circle. It would be good to explain this stuff more clearly, and include some "Do x and you'll observe y" type explanation after the demo. That is a lot of code to dump onto the page without any explanation interspersed.

Also:

- The three form controls look messy because one of them has fallen down to the line below the other two (on Firefox at least). Put them all on separate lines?
- It feels like the example is a bit big, and work just as well if it were made smaller.
- As a consequence of the example size and the fact that it scrolls, I missed the `clip-path` property value output printed below the "canvas" at the bottom. It is good that you are doing this (I was feeling pretty insensed that you hadn't provided that, until I found it at the bottom of the demo, super useful). I think you should put it in a more prominent place, like maybe overlaid on top of the "canvas" itself.

Author: Thanks @chrisdavidmills - I have accepted comments

Rewored the example to have sub headings, in-place description. Moved the clip-path value description below the controls so you can't miss it now. The controls draw 2 per line, which will work on most screens. I chose not to put one on each line because you already felt the example was too big.

I'm not reducing the size of the example - it is just right. It looks small if you use the closest-X options, but can easily overflow and clip on the sides if you set a farthest corner.

To confirm one point, the reference box contains a fill box, which defines a gradient (way) outside of the reference box. The properties of the clip path are defined on the reference box but that doesn't mean that the clip is constrained the ref box. 
The way clip works it that it can clip anything that the element paints, including its contents. So the fact that content draws outside the reference box is why the clip also has effect outside the ref box.

{{EmbedLiveSample("circle", "", "300px")}}

> ### Radius keyword interactive example

Author: FYI A lot of this is claude generated. It makes sense, but depending on reviewer thoughts perhaps it might be simplified.

  gap: 20px;
  align-items: center;
}

.canvas {
  position: relative;
  width: 640px;
  height: 480px;
  overflow: hidden;
  border: 1px solid #888;
}

.outline,
.refbox {
  position: absolute;
  top: 180px;
  left: 220px;
  width: 200px;
  height: 120px;
}

.outline {
  border: 3px dashed #e6007a;
  pointer-events: none;
}

.fill {
  position: absolute;
  inset: -240px;
  background: linear-gradient(to bottom right, #ff5522, #0055ff);
}

.center-marker {
  position: absolute;
  width: 10px;
  height: 10px;
  border-radius: 50%;
  background: white;
  border: 2px solid black;
  transform: translate(-50%, -50%);
  pointer-events: none;
}

#support-note {
  color: #b3001b;
}
```

```js live-sample___circle-keywords-interactive
const select = document.getElementById("radius-keyword");
const posX = document.getElementById("pos-x");
const posY = document.getElementById("pos-y");
const refbox = document.getElementById("refbox");
const declaration = document.getElementById("declaration");
const marker = document.getElementById("center-marker");
const supportNote = document.getElementById("support-note");

// Feature-test each keyword, since closest-corner/farthest-corner are a

> // newer addition to circle() and may not be supported everywhere yet.

Author: ```suggestion
// Feature-test each keyword
```

  </div>
</div>

<pre id="declaration"></pre>

<p id="support-note"></p>
```

The HTML then defines a number of {{htmlelement("div")}} elements that are used to render the circle, reference box and center marker.
The most important elements are the reference box (`#refbox`) and the `.fill` element it contains, which define the gradient that is drawn (as discussed in the following CSS section).
A `clipPath` is set on the reference box in JavaScript to clip that gradient to a circular shape.

```html live-sample___circle-keywords-interactive
<div class="canvas">
  <div class="refbox" id="refbox">
    <div class="fill"></div>
  </div>
  <div class="outline"></div>
  <div class="center-marker" id="center-marker"></div>
</div>
```

Note that the `.outline` and `.center-marker` are provided to make it easier to see the reference box and the center of the clipping circle, which would otherwise be invisible.
The `.outline` element must be placed after `#refbox` in the markup: since it isn't clipped itself, painting it after the reference box means its dashed border always renders on top of the `.fill` gradient, even when the clipping circle extends beyond the reference box.

#### CSS

> The CSS for the canvas, reference box and fill elements are shown below.

Reviewer: ```suggestion
The CSS for the canvas, reference box, and fill elements is shown below.
```

  box-sizing: border-box;
}

.fill {
  position: absolute;
  inset: -240px;
  background: linear-gradient(to bottom right, #ff5522, #0055ff);
}
```

The CSS for the controls and other elements is not shown, as it isn't needed to understand the `circle()` method.
Interested readers can view it in the example playground.

```css live-sample___circle-keywords-interactive hidden
body {
  display: flex;
  flex-direction: column;
  gap: 20px;
  align-items: flex-start;
}

.controls {
  display: flex;
  flex-direction: column;
  gap: 20px;
}

.controls-row {
  display: flex;
  flex-wrap: wrap;
  gap: 20px;
  align-items: center;
}
```

```css live-sample___circle-keywords-interactive hidden
.outline {
  position: absolute;
  top: 180px;
  left: 220px;
  width: 200px;
  height: 120px;
  box-sizing: border-box;
}

.outline {
  border: 3px dashed #e6007a;
  pointer-events: none;
}

.center-marker {
  position: absolute;
  width: 10px;
  height: 10px;
  border-radius: 50%;
  background: white;
  border: 2px solid black;
  transform: translate(-50%, -50%);
  pointer-events: none;
}

#declaration {
  margin: 0;
}

#support-note {
  margin: 0;
  color: #b3001b;
}
```

#### JavaScript

> The JavaScript for the example is shown below.

Reviewer: ```suggestion
```

I don't think this line is needed.

> First we get handles to each of the elements used by the example.

Reviewer: You could put this line above a code block just featuring the const declarations, and then put the "Then we define..." bit above a code block just featuring the `checkSupport()` function definition.

> Then we define `checkSupport()` method to feature-test whether each radius keyword is supported and enable/disable the associated selection options.

Reviewer: ```suggestion
Then we define the `checkSupport()` function used to feature-test whether each radius keyword is supported and enable/disable the associated selection options.
```

  }
  if (select.selectedOptions[0]?.disabled) {
    const firstSupported = [...select.options].find(
      (option) => !option.disabled,
    );
    if (firstSupported) {
      select.value = firstSupported.value;
    }
  }
}
```

Next we define an `update()` function that is called whenever the input controls are changed.
This sets the clip path on the reference box based on the selected keyword and position, and moves the center marker to match.

```js live-sample___circle-keywords-interactive
function update() {
  const keyword = select.value;
  const x = Number(posX.value);
  const y = Number(posY.value);

  // Build the clip-path value and apply it to the reference box
  const value = `circle(${keyword} at ${x}px ${y}px)`;
  refbox.style.clipPath = value;

  // Update the displayed declaration text and marker position
  declaration.textContent = `clip-path: ${value};`;
  marker.style.left = `${refbox.offsetLeft + x}px`;
  marker.style.top = `${refbox.offsetTop + y}px`;
}

select.addEventListener("change", update);
posX.addEventListener("input", update);
posY.addEventListener("input", update);
checkSupport();
update();
```

#### Result

> Change the position of the center and the keyword used to see their relative effects.

Reviewer: ```suggestion
Change the position of the circle center and the keyword used to see their relative effects.
```
